app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,10 +113,31 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    hashes_index = {}
+    hashes.each do |hash|
+      hashes_index[build_index_from_hash(keys, hash)] = hash
+    end
+
+    records.find_each do |record|
```

Comment:
the records are preloaded, but duplicated, so I rather fetch them again in batches